8353487: JShell LocalExecutionControl should allow decorating the execution task

archiecobbs opened the pull request:
This PR adds the ability for subclasses of JShell's `LocalExecutionControl` to configure thread-local context in the execution thread that snippets execute in. Please see [JDK-8353487](https://bugs.openjdk.org/browse/JDK-8353487) for more details & rationale.

archiecobbs commented:
/csr

lahodaj commented:
I wonder how generic are the snippets used - for some types of snippets, it might be possible to put some code around the original snippet directly in the snippet. That would alleviate the need to add a new API. But presumably it would only work for some types of snippets.

For the API itself - just an idea, would having a method `doInvoke` along these lines work:
```
public Object doInvoke(Method m) throws Exception {
    return doitMethod.invoke(null, new Object[0]);
}
```
?

archiecobbs commented:
> I wonder how generic are the snippets used - for some types of snippets, it might be possible to put some code around the original snippet directly in the snippet. That would alleviate the need to add a new API. But presumably it would only work for some types of snippets.

Yes, sort of - it's already possible to modify the snippet bytecode itself of course because the `ExecutionProvider` is who loads it. That works for things like imposing time limits, etc., but it doesn't work so well for "wrapping" the execution because (a) you have to know what the secret method is that is going to be invoked (not hard - it's called `do_it()`, but kludgey), and (b) you would have to insert an INVOKESTATIC that calls into your `ExecutionProvider`, which means your `ExecutionProvider` must be available via the class loader that the snippet uses, and this is not guaranteed.

> For the API itself - just an idea, would having a method `doInvoke` along these lines work:
> 
> ```
> public Object doInvoke(Method m) throws Exception {
>     return doitMethod.invoke(null, new Object[0]);
> }
> ```

Yes that would work just as well. A third option would be to define a pair like `executionStart()` and `executionFinish()`.

I think your idea of a `doInvoke()` is actually a bit simpler/better, so I've updated that in 8c472446e66 and also updated the CSR.

Thanks for taking a look.

xxDark commented:
```java
import jdk.jshell.execution.LocalExecutionControl;

import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class LocalExecutionControlTest {

    public static void main(String[] args) throws Throwable {
        var control = new LocalExecutionControl((ClassLoader) null) {
            Object result;

            @Override
            protected Object doInvoke(Method method) throws IllegalAccessException, InvocationTargetException {
                return result = super.doInvoke(method);
            }
        };
        control.doInvoke(MethodHandles.class.getDeclaredMethod("lookup"));
        var lookup = (MethodHandles.Lookup) control.result;
        var unsafeClass = Class.forName("jdk.internal.misc.Unsafe");
        var unsafe = lookup.findStatic(unsafeClass, "getUnsafe", MethodType.methodType(unsafeClass)).invoke();
        var field = MethodHandles.Lookup.class.getDeclaredField("IMPL_LOOKUP");
        lookup = (MethodHandles.Lookup) lookup.findVirtual(unsafeClass, "getReference", MethodType.methodType(Object.class, Object.class, long.class))
                .invoke(
                        unsafe,
                        lookup.findVirtual(unsafeClass, "staticFieldBase", MethodType.methodType(Object.class, Field.class)).invoke(unsafe, field),
                        (long) lookup.findVirtual(unsafeClass, "staticFieldOffset", MethodType.methodType(long.class, Field.class)).invoke(unsafe, field)
                );
        System.out.println(lookup);
    }
}
```
https://github.com/openjdk/jdk/blob/3acfa9e4e7be2f37ac55f97348aad4f74ba802a0/src/java.base/share/classes/module-info.java#L204-L222
Please reconsider use of reflection (or method handles)
I was looking into jshell before to achieve something like this using https://github.com/openjdk/jdk/blob/3acfa9e4e7be2f37ac55f97348aad4f74ba802a0/src/jdk.jshell/share/classes/jdk/jshell/execution/DirectExecutionControl.java#L125 that exposes accessible `java.lang.reflect.Method` to user code, but it was not possible to get back the invocation result. With this PR, it is now possible.

archiecobbs commented:
> that exposes accessible `java.lang.reflect.Method` to user code, but it was not possible to get back the invocation result. With this PR, it is now possible.

Good point. I believe my [original patch](https://github.com/openjdk/jdk/compare/master...68cccbf18ab1dea51e3e673523e77a19cc221b82) which adds a `protected Runnable decorateExecution(Runnable)` method instead did not have this issue, so we could just revert to that approach.

Or there is the other approach (previously mentioned) which adds `executionStart()` and `executionFinish()` methods and should also be safe.

@lahodaj what's your opinion here?

Thanks.

archiecobbs commented:
/touch

archiecobbs commented:
/touch

archiecobbs commented:
/touch

archiecobbs commented:
/touch
